### PR TITLE
Allows pushing page based route as pageless route

### PR DESCRIPTION
**Pull request by chunhtai**
related https://github.com/flutter/flutter/issues/114216#issuecomment-1295265318

The use case is to be able to reuse routing table for imperative API. One common feature is user want to push a certain page in the route table on top of current page instead of reseting the entire route stack.

Currently go_router use a hack to modify the internal state to display the push page and potentially make internal state and url out of sync.

The goal is that instead of mutating the internal state, it uses the naive navigator API to push a pagebased route from the route match as a pageless route.

## Pre-launch Checklist

- [ ] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [ ] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [ ] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [ ] I signed the [CLA].
- [ ] I listed at least one issue that this PR fixes in the description above.
- [ ] I updated/added relevant documentation (doc comments with `///`).
- [ ] I added new tests to check the change I am making, or this PR is [test-exempt].
- [ ] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

**Comment by goderbauer** [comment]
Can you check out the failures?

File: packages/flutter/lib/src/widgets/navigator.dart [hunk 1]
```diff
@@ -3715,14 +3718,11 @@ class NavigatorState extends State<Navigator> with TickerProviderStateMixin, Res
       );
       // Pageless routes will be recorded when we update the middle of the old
       // list.
-      if (!oldEntry.hasPage) {
+      if (!oldEntry.pageBased) {
         continue;
       }
 
-      assert(oldEntry.hasPage);
```
Comment (goderbauer) [inline packages/flutter/lib/src/widgets/navigator.dart:3722]: Why remove this?

**Review by goderbauer — APPROVED** [review]
LGTM

File: packages/flutter/lib/src/widgets/navigator.dart (hunk 1 above)
Comment (chunhtai) [reply packages/flutter/lib/src/widgets/navigator.dart:3722]: I think this check is redundant  given the if case above should already short-circuited pageless route

**Comment by chunhtai** [comment]
google test is somehow buggy I am going to manually merge this

**Comment by JunioJsv** [comment]
is this some sort of nested route? I've been looking for a long time for a way to pass my dependencies between routes without passing them as a parameter in the constructor or reinserting them when generating the new route.

**Comment by JunioJsv** [comment]
is this some sort of nested route? I've been looking for a long time for a way to pass my dependencies between routes without passing them as a parameter in the constructor or reinserting them when generating the new route.

**Comment by chunhtai** [comment]
@JunioJsv no, this is not related to nested route

**Pull request merged**